### PR TITLE
[Merged by Bors] - feat(Algebra/Category): for ring homs, epi + finite => surjective

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/Ring/Epi.lean [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    exact RingHom.congr_fun this
+  · refine fun H ↦ ⟨fun {T} f g e ↦ ?_⟩
+    letI : Algebra R T := (ofHom (algebraMap R S) ≫ g).toAlgebra
+    let f' : S →ₐ[R] T := ⟨f, RingHom.congr_fun e⟩
+    let g' : S →ₐ[R] T := ⟨g, fun _ ↦ rfl⟩
+    ext s
+    simpa using congr(Algebra.TensorProduct.lift f' g' (fun _ _ ↦ .all _ _) $(H s))
+
+lemma RingHom.surjective_of_epi_of_finite {R S : CommRingCat} (f : R ⟶ S) [Epi f]
+    (h₂ : RingHom.Finite f) : Function.Surjective f := by
```
Comment (chrisflav) [inline Mathlib/Algebra/Category/Ring/Epi.lean:39]: Maybe also add the `iff` version?

**Review by chrisflav — APPROVED** [review]
Nice result!

File: Mathlib/Algebra/Category/Ring/Epi.lean [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  letI := f.toAlgebra
+  have : Module.Finite R S := h₂
```
Comment (callesonne) [inline Mathlib/Algebra/Category/Ring/Epi.lean:42]:
```suggestion
  algebraize [f]
```
Does this work?

Comment (erdOne) [reply Mathlib/Algebra/Category/Ring/Epi.lean:42]: No it complains that `f : R ⟶ S` is not a `RingHom`.

Comment (callesonne) [reply Mathlib/Algebra/Category/Ring/Epi.lean:42]: Oooooh, I should definitely fix that at some point!

File: Mathlib/RingTheory/TensorProduct/Basic.lean [hunk 3]
```diff
@@ -7,6 +7,8 @@ import Mathlib.Algebra.Algebra.RestrictScalars
 import Mathlib.GroupTheory.MonoidLocalization.Basic
 import Mathlib.LinearAlgebra.TensorProduct.Tower
 import Mathlib.RingTheory.Adjoin.Basic
+import Mathlib.RingTheory.Finiteness
+import Mathlib.Algebra.Algebra.RestrictScalars
```
Comment (chrisflav) [inline Mathlib/RingTheory/TensorProduct/Basic.lean:11]:
```suggestion
```
this is imported twice

**Review by chrisflav — APPROVED** [review]

**Comment by chrisflav** [comment]
Otherwise, LGTM

File: Mathlib/RingTheory/TensorProduct/Basic.lean [hunk 4]
```diff
@@ -7,6 +7,8 @@ import Mathlib.Algebra.Algebra.RestrictScalars
 import Mathlib.GroupTheory.MonoidLocalization.Basic
 import Mathlib.LinearAlgebra.TensorProduct.Tower
 import Mathlib.RingTheory.Adjoin.Basic
+import Mathlib.RingTheory.Finiteness
```
Comment (chrisflav) [inline Mathlib/RingTheory/TensorProduct/Basic.lean:10]: This basic file was recently split, so you can just move your lemma here to `Mathlib.RingTheory.TensorProduct.Finite`.

**Comment by chrisflav** [comment]
Thanks!

maintainer merge

**Pull request closed without merging**